boxtree/tree_build.py

```diff
@@ -37,6 +37,242 @@ class MaxLevelsExceeded(RuntimeError):
+        lev = self.box_levels[ibox]
+        box_size = self.root_box_extent * 0.5**lev
+        return box_size
+
+    def get_box_extent(self, ibox):
+        box_size = self.get_box_size(ibox)
+        extent_low = self.box_centers[:, ibox] - 0.5*box_size
+        extent_high = extent_low + box_size
+        return extent_low, extent_high
+
+    # }}} End dummy interface for TreePlotter
+
+    @property
+    def nboxes(self):
+        return self.box_centers.shape[1]
+
+    def nlevels(self):
+        return max(self.box_levels) + 1  # level starts from 0
+
+    def is_leaf(self):
+        # box_id -> whether the box is leaf
+        return np.all(self.box_children == 0, axis=0)
+
+    def leaf_boxes(self):
+        boxes = np.arange(self.nboxes)
+        return boxes[self.is_leaf()]
+
+    # {{{ instantiate with particles
+
+    def distribute_quadrature_rule(self, quad_rule):
+        """Generate quadrature nodes (`dim x nnodes`) and weights. Within each
+        leaf box, nodes are in lexicographic order, where the last axis varies
+        fastest. For example, `[[0, 0, 1, 1], [0, 1, 0, 1]]`.
+
+        :param quad_rule: a :class:`modepy.Quadrature`
+        """
+        x, w = quad_rule.nodes, quad_rule.weights  # nodes in [-1, 1]
+        n_box_nodes = len(x)**self.dim
+        box_nodes = np.array(np.meshgrid(*((x,) * self.dim), indexing="ij")
+                             ).reshape([self.dim, -1])
+        if self.dim == 2:
+            box_weights = w[:, None] @ w[None, :]
+        elif self.dim == 3:
+            box_weights = w[:, None, None] @ w[None, :, None] @ w[None, None, :]
+        box_weights = box_weights.reshape(-1)
+        lfboxes = self.leaf_boxes()
+
+        nodes = np.tile(box_nodes, (1, len(lfboxes)))
+        box_shifts = np.repeat(
+            self.box_centers[:, lfboxes], n_box_nodes, axis=1)
+        box_scales = np.repeat(
+            self.get_box_size(lfboxes) / 2, n_box_nodes)
+        nodes = nodes * box_scales[None, :] + box_shifts
+
+        weights = np.tile(box_weights, len(lfboxes))
+        weights = weights * box_scales**self.dim
+
+        return nodes, weights
+
+    # }}} End instantiate with particles
+
+    # {{{ refine/coarsen
+
+    def refined(self, refine_flags):
+        return self.refine_and_coarsened(refine_flags, None)
+
+    def uniformly_refined(self):
+        refine_flags = np.zeros(self.nboxes, bool)
+        refine_flags[self.is_leaf()] = 1
+        return self.refined(refine_flags)
+
+    def coarsened(self, coarsen_flags):
+        return self.refine_and_coarsened(None, coarsen_flags)
+
+    def refine_and_coarsened(self, refine_flags=None, coarsen_flags=None):
+        """Make a refined/coarsened copy. When children of the same parent box
+        are marked differently, the refinement flag takes priority.
+
+        Both refinement and coarsening flags can only be set of leafs.
+        To prevent drastic mesh change, coarsening is only executed when a leaf
```

Comment:
What is "drastic"? (define or omit)

Reply:
95c170f34a263c693842acf6730cc8499ccdc1dd